protos/core.proto

```diff
@@ -247,6 +247,190 @@ message NotebookRuntimeOptions {
   }
 }
 
+// Data Preparation Related entries
+message DataPreparation {
+  // Data preparatiohs can have more than 1 output
+  repeated Target targets = 1;
+
+  repeated Target canonical_targets = 2;
+
+  repeated string tags = 3;
+
+  repeated Target dependency_targets = 4;
+
+  string file_name = 5;
+
+  bool disabled = 6;
+
+  DataPreparationDefinition data_preparation = 7;
+}
+
+message DataPreparationDefinition {
+  repeated DataPreparationNode nodes = 1;
+  DataPreparationGenerated generated = 2;
+}
```

Comment:
I don't really understand why this is structured in this way - with nested nodes.

There are many downsides to having a node that contains other nested nodes:
* What prevents it from becoming a cyclic dag?
* How would one represent this in a UI?

It would be much better to keep each node as separate actions instead.

Reply:
The DAG may contain branches or nodes that join 2 other nodes. The DAG is structured in a way that a node always points into a Source:

- A BigQuery Table
- 1 previous node
- 2 previous nodes, which are combined using a SQL JOIN operation